### ISSUE TITLE
Encoded characters are not correctly counted for node size

### ISSUE THREAD

**Issue by azaroth42** [knsv/mermaid#975]
**Describe the bug**

When using special characters in encoded form, such as `#quot;` the resulting box is too wide. It clearly counts all of the characters in the encoded form, but then renders the character ... without shrinking the size of the box.

**To Reproduce**

* Create a diagram that uses `#quot;`s inside a node, such as:

```
graph TD
A["#quot;Christmas#quot;"] -->|Get money| B(Go shopping)
```

Or in the [editor](https://mermaidjs.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbkFbXCIjcXVvdDtDaHJpc3RtYXMjcXVvdDtcIl0gLS0-fEdldCBtb25leXwgQihHbyBzaG9wcGluZylcblxuIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifX0)

**Expected behavior**

`#quot;` should be counted as only one character for determining the width of the node.

**Screenshots**

![Screen Shot 2019-10-05 at 9 43 33 PM](https://user-images.githubusercontent.com/871868/66260662-49861200-e7b9-11e9-8407-337e4b97c39f.png)

**Comment by sukhbeersingh** [comment]
Hi, Id like to take a look into this. Can you direct me how to start?

**Comment by Dunning-Kruger** [comment]
Hey, @sukhbeersingh. Feel free to drop by our slack [here](https://join.slack.com/t/mermaid-talk/shared_invite/enQtNzc4NDIyNzk4OTAyLWVhYjQxOTI2OTg4YmE1ZmJkY2Y4MTU3ODliYmIwOTY3NDJlYjA0YjIyZTdkMDMyZTUwOGI0NjEzYmEwODcwOTE) and get in contact with some other developers of the libary!

**Comment by knsv** [comment]
@sukhbeersingh Here comes a small direction. flowRenderer in the addVeritces function is probably involved in this. It seems the sizing is done from text instead of the actual rendered size.

If the issue is when rendering without htmlLabels things might be easier, then I think you can convert them back. We have to be careful about not exposing ourselves for xss attacks. There is an e2e test for that.

If you run cypress open you can trigger the xss tests and check. The file with these tests are xss.spec.js.

Feel free to add more of them if it helps. We probably should. With and without htmlLabels are probably a good idea.

For the xss tests you can add options to the rendering in the second argument to mermaidUrl function. Thats where you switch between different security setting and switch htmlLabels on/off. (More info about the options in the mermaidAPI.md/mermaidAPI.js.

```
    const url = mermaidUrl(str,{}, true);
```

I suspect that this is either fairly easy or quite tricky. So don't feel bad if it turns out to be tricky. Worst case it will need to be postponed until https://github.com/knsv/mermaid/projects/4.

**Comment by sukhbeersingh** [comment]
Thanks for the info. htmlLabels are indeed being used to render this kind of text and then it is being decoded by decodeEntities, but by then the size is decided for the box. What if we adjust the size of the shape after it has been decoded from the actual text itself? 

Also, can we not parse the text for its actual size before sending to addHtmlLabel function?

**Comment by jgreywolf** [comment]
@sukhbeersingh Are you still working on this?  Any update?

**Comment by bebuch** [comment]
On our GitLab instance we are also still waiting on this.

**Comment by sukhbeersingh** [comment]
Hi, since I didn't get a reply back on how to approach the question, I didnt progress further. You can read it in my blog: https://sukhbeerdhillon.medium.com/a-story-of-failure-and-learning-24a4e766d1e9

**Comment by imperator-maximus** [comment]
workaround for #quot;: just just use ” instead of " or  #quot;

**Comment by GiovanH** [comment]
Looking at the example provided in the live editor shows that this has been fixed already.
